src/llm_tracekit/bedrock/invoke_agent.py

```diff
@@ -87,78 +192,166 @@ def record_invoke_agent_result_attributes(
         record_metrics(
             instruments=instruments,
             duration=duration,
-            usage_input_tokens=usage_input_tokens,
-            usage_output_tokens=usage_output_tokens,
+            usage_input_tokens=result.usage_input_tokens,
+            usage_output_tokens=result.usage_output_tokens,
+            response_model=result.foundation_model,
         )
 
 
 class InvokeAgentStreamWrapper(ObjectProxy):
-    """Wrapper for botocore.eventstream.EventStream"""
+    """
+    A wrapper for botocore.eventstream.EventStream that intercepts and processes
+    events from the Bedrock `invoke_agent` API to gather telemetry data.
+    It is designed to work robustly whether `enableTrace` is True or False.
+    """
 
     def __init__(
         self,
         stream: EventStream,
-        stream_done_callback: Callable[
-            [Optional[str], Optional[int], Optional[int]], None
-        ],
+        stream_done_callback: Callable[[AgentStreamResult], None],
         stream_error_callback: Callable[[Exception], None],
     ):
         super().__init__(stream)
-
         self._stream_done_callback = stream_done_callback
         self._stream_error_callback = stream_error_callback
-        self._content = None
+
+        self._content: Optional[str] = None
         self._usage_input_tokens: Optional[int] = None
         self._usage_output_tokens: Optional[int] = None
+        self._foundation_model: Optional[str] = None
+        self._inference_config_max_tokens: Optional[int] = None
+        self._inference_config_temperature: Optional[float] = None
+        self._inference_config_top_k: Optional[int] = None
+        self._inference_config_top_p: Optional[float] = None
+        self._chat_history: Optional[Tuple[List[Message], List[Choice]]] = None
+        self._tool_calls_buffer: List[ToolCall] = []
```

Comment:
Since you just put all of these in a `AgentStreamResult` object at the end of iteration, consider replacing all of these with an internal `self._result = AgentStreamResult()`, then modify the fields inside of it when iterating over the stream (e.g. doing `self._result.usage_prompt_tokens += usage.get("inputTokens", 0)` instead of `self._usage_prompt_tokens += usage.get("inputTokens", 0)`)